src/moaicore/MOAICCParticleSystem.cpp

```diff
@@ -0,0 +1,2377 @@
+        doc.LoadFile ( xml );
+        self->ParseXML ( xml, doc.RootElement ());
+        
+    }
+    
+    return 0;
+}
+//----------------------------------------------------------------//
+/** @name   startSystem
+    @text   Begin the generation of particles.  Use in conjenction with start().
+ 
+    @in     MOAICCParticleSystem    self
+    @out    nil
+ */
+
+int MOAICCParticleSystem::_startSystem ( lua_State *L ) {
+    MOAI_LUA_SETUP( MOAICCParticleSystem, "U" )
+    self->StartSystem();
+    return 0;
+}
+
+//----------------------------------------------------------------//
+/** @name   stopSystem
+    @text   Halt the generation of particles.  Any remaining particles will be allowed to stay until the end of their lifespan. Use stop() to halt the animation of particles.
+ 
+    @in     MOAICCParticleSystem    self
+    @out    nil
+ */
+int MOAICCParticleSystem::_stopSystem ( lua_State *L ) {
+    
+    MOAI_LUA_SETUP( MOAICCParticleSystem, "U" )
+    self->StopSystem();
+    
+    return 0;
+}
+//----------------------------------------------------------------//
+/** @name   surge
+    @text   Generate the given number of particles at one time.
+    @in     MOAICCParticleSystem    self
+    @in     number                  numberOfParticles
+    @out    nil
+ */
+
+int MOAICCParticleSystem::_surge ( lua_State *L ) {
+    MOAI_LUA_SETUP( MOAICCParticleSystem, "UN" )
+    u32 numberOfParticles = state.GetValue < u32 >(2, 1.0);
+    self->Surge( numberOfParticles );
+    return 0;
+}
+//----------------------------------------------------------------//
+/** @name   reset
+    @text   Make all particles get cleared out at the next frame.
+ 
+    @in     MOAICCParticleSystem    self
+    @out    nil
+ */
+
+int MOAICCParticleSystem::_reset ( lua_State *L ) {
+    MOAI_LUA_SETUP( MOAICCParticleSystem, "U" )
+    self->ResetSystem();
+    return 0;
+}
+
+//----------------------------------------------------------------//
+//----------------------------------------------------------------//
+
+bool MOAICCParticleSystem::AddParticle () {
+    if ( this->IsFull() ) {
+        return false;
+    }
+    
+    MOAICCParticle particle(this);
+    this->mParticles.push_back(particle);
+    
+    this->mParticleCount++;
+    
+    return true;
+}
+
+void MOAICCParticleSystem::Draw ( int subPrimID ) {
+    UNUSED(subPrimID);
+    
+    if ( !this->mDeck ) {
+        return;
+    }
+    
+    MOAIGfxDevice& gfxDevice = MOAIGfxDevice::Get ();
+    
+    if ( this->mUVTransform ) {
+        USAffine3D uvMtx = this->mUVTransform->GetLocalToWorldMtx ();
+        gfxDevice.SetUVTransform ( uvMtx );
+    }
+    else {
+        gfxDevice.SetUVTransform ();
+    }
+    
+    this->LoadGfxState();
+    
+    USAffine3D drawingMtx;
+    USAffine3D spriteMtx;
+    
+    for (auto particleIterator = this->mParticles.begin(), end = this->mParticles.end() ; particleIterator != end; ++particleIterator) {
+        MOAICCParticle *particle = &(*particleIterator);
```

Comment:
There's an even shorter version of this:

    for (auto particle: this->mParticles)